Add rwsem_tryupgrade for 4.9.20-rt16 kernel

behlendorf opened the pull request:
### Description

The RT rwsem implementation was changed to allow multiple readers
as of the 4.9.20-rt16 patch set.  This results in a build failure
because the existing implementation was forced to directly access
the rwsem strucgture which has changed.

While this could be accommodated by adding additional compatibility
code.  This patch resolves the build issue by simply assuming the
rwsem can never be upgraded.  This functionality is a performance
optimization and all callers must already handle this case.

### Motivation and Context

zfsonlinux/spl#611

### How Has This Been Tested?

Locally built, which isn't saying much.  But it should work @clefru, @kernelOfTruth, or @steven-omaha would you be able to test that this resolves the build issue.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ ] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).
- [ ] Change has been approved by a ZFS on Linux member.

steven-omaha commented:
Awesome! Thanks for working on this. I will building it on one of my machines next week.

steven-omaha commented:
@behlendorf I checked your patch on one of my machines, but it didn't work for me. Maybe I didn't do the right things, but I'll try to explain as best as I can. Sorry if I did something stupid, this is my first time fiddling around with these things.

This is on Arch Linux. Here ZFS is a user-maintained package that is actually divided into 4 packages `spl-dkms, spl-utils, zfs-dkms, zfs-utils`. I used the latest stable release for spl (0.7.9), and applied your patch, (see the build-file here)[https://pastebin.com/5ykYMJ0v]. Specifically, I added the lines 15 and 22 to fetch and apply your patch. I checked to build directory to verify that the patch as actually applied.

Building the package works, and building the dkms-modules for the vanilla-, lts- and zen-kernel works, but not for the rt-kernel.

`sudo dkms install spl/0.7.9 -k 4.16.8-rt3-1-rt` leads to the output [here](https://pastebin.com/vR4LYz3g). (/var/lib/dkms/spl/0.7.9/build/make.log)

I can make further tests, just tell me.

behlendorf commented:
@steven-omaha thanks for testing.  When I get a chance, I'll give this a test run with the Arch Linux RT kernel which looks like a convenient way to test this.

behlendorf commented:
@steven-omaha I've updated this PR and addressed the issue you likely encountered.  Without this change I was able to reproduce the reported build issue when using the Debian `4.9.0-7-rt-amd64` kernel.  With the change applied I'm able to build cleanly.  Sorry about the delay.

steven-omaha commented:
@behlendorf Thanks for working on this. I tried to build it, but to no avail. I think the error is on my side. I'm not really familiar with building projects as complex as zfs / spl. I'm using the same building steps again (see above), but I'm now patching the source using your commit a38d724. It doesn't matter if I try this from the last tagged version, or the recent master. I always get failed patching.

    patching file /tmp/makepkg/spl-dkms/src/spl/module/spl/spl-rwlock.c
    Hunk #1 FAILED at 172.
    Hunk #2 FAILED at 188.
    2 out of 2 hunks FAILED -- saving rejects to file /tmp/makepkg/spl-dkms/src/spl/module/spl/spl-rwlock.c.rej
    patching file /tmp/makepkg/spl-dkms/src/spl/module/spl/spl-rwlock.c
    Hunk #1 succeeded at 34 (offset -1 lines).
    Hunk #2 succeeded at 67 (offset -1 lines).
    patching file /tmp/makepkg/spl-dkms/src/spl/module/spl/spl-rwlock.c
    Hunk #1 FAILED at 744.
    1 out of 1 hunk FAILED -- saving rejects to file /tmp/makepkg/spl-dkms/src/spl/module/spl/spl-rwlock.c.rej`

behlendorf commented:
@steven-omaha try this.  Instead of wrangling the patch and manually applying it you can download the pre-patched source as a tarball from github like this. 

```
curl https://codeload.github.com/zfsonlinux/zfs/legacy.tar.gz/a38d724 | tar -xz
cd zfsonlinux-zfs-a38d724
sh autogen.sh
./configure
make
```

steven-omaha commented:
@behlendorf I checked this on Arch using `4.16.18-rt9`, and I could compile it using your steps.
